instrumentation/rack/lib/opentelemetry/instrumentation/rack/middlewares/event_handler.rb

```diff
@@ -0,0 +1,255 @@
+              unless request_start_time.nil?
+                frontend_span = @tracer.start_span(
+                  'http_server.proxy',
+                  with_parent: extracted_context,
+                  start_timestamp: request_start_time,
+                  kind: :server
+                )
+
+                frontend_context = OpenTelemetry::Trace.context_with_span(frontend_span, parent_context: extracted_context)
+              end
+            end
+            parent_context = frontend_context || extracted_context
+
+            span = @tracer.start_span(
+              create_request_span_name(request),
+              with_parent: parent_context,
+              kind: frontend_context.nil? ? :server : :internal,
+              attributes: request_span_attributes(request.env)
+            )
+            ctx = OpenTelemetry::Trace.context_with_span(span)
+            rack_ctx = OpenTelemetry::Instrumentation::Rack.context_with_span(span, parent_context: ctx)
+
+            contexts = [frontend_context, ctx, rack_ctx]
+            contexts.compact!
+
+            tokens = contexts.map { |context| OpenTelemetry::Context.attach(context) }
+
+            request.env[TOKENS_KEY] = tokens
+          end
+
+          # Optionally adds debugging response headers injected from {response_propagators}
+          #
+          # @param [Rack::Request] The current HTTP request
+          # @param [Rack::Response] This current HTTP response
+          # @return [void]
+          def on_commit(request, response)
+            span = OpenTelemetry::Instrumentation::Rack.current_span
+            return unless span.recording?
+
+            begin
+              @response_propagators&.each { |propagator| propagator.inject(response.headers) }
+            rescue StandardError => e
+              OpenTelemetry.handle_error(message: 'Unable to inject response propagation headers', exception: e)
+            end
+          end
+
+          # Records Unexpected Exceptions on the Rack span and set the Span Status to Error
+          #
+          # @note does nothing if the span is a non-recording span
+          # @param [Rack::Request] The current HTTP request
+          # @param [Rack::Response] The current HTTP response
+          # @param [Exception] An unxpected error raised by the application
+          def on_error(request, _, error)
+            span = OpenTelemetry::Instrumentation::Rack.current_span
+            return unless span.recording?
+
+            span.record_exception(error)
+            span.status = OpenTelemetry::Trace::Status.error
+          end
+
+          # Finishes the span making it eligible to be exported and cleans up existing contexts
+          #
+          # @note does nothing if the span is a non-recording span
+          # @param [Rack::Request] The current HTTP request
+          # @param [Rack::Response] The current HTTP response
+          def on_finish(request, response)
```

Comment:
It might also be nice to add `on_send` and record a span event for "sent first byte" or something similar. It is really nice to know, from a perf analysis perspective, when the application has finalized the response and the remaining work is "out of our hands".

An interesting philosophical question: when should we consider the server span "finished"? When application processing is complete or when the response has been completely written to the socket? I _think_ the current middleware implementation is effectively equivalent to "when application processing is complete", before anything has been written to the socket.

Reply:
> It might also be nice to add `on_send` and record a span event for "sent first byte" or something similar. It is really nice to know, from a perf analysis perspective, when the application has finalized the response and the remaining work is "out of our hands".
> 

I like this idea a lot! I will add it to my todo list.

> 
> An interesting philosophical question: when should we consider the server span "finished"? When application processing is complete or when the response has been completely written to the socket? I _think_ the current middleware implementation is effectively equivalent to "when application processing is complete", before anything has been written to the socket.

- 💭 I'm not sure. I lean towards after the middleware has completed writing the response but does that make it asymmetric from when it's processing the incoming request? I'm not entirely certain.

Reply:
Going to hold off adding other events until I get this initial implementation merged. 

I am also going to stick with `on_finish`, i.e. after the response has been written to the socket and the `Rack::BufferedResponse` is closed.